{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "968e7594-39c0-4e5f-9e3c-43f022dfd6ec",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done.\n"
     ]
    }
   ],
   "source": [
    "#mandatory imports\n",
    "import time\n",
    "from Bio import Entrez\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import spacy\n",
    "import urllib.request, urllib.error, urllib.parse\n",
    "import json\n",
    "print(\"Done.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "13c8df16-9197-41d2-9e4a-62e8795c3c93",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done.\n"
     ]
    }
   ],
   "source": [
    "## Thesaurus declaration\n",
    "thesaurus = {\n",
    "    \"Ivermectin\" : [\"IVM\", \"Ivermectin\", \"Avermectin\", \"avermectin\", \"ivermectin\", \"stromectol\", \"Stromectol\", \"Eqvalan\", \"Ivomec\", \"Mectizan\", \"Dihydroavermectin\", \"MK 933\", \"MK-933\", \"MK933\",\"MK-0933\", \"L 640471\", \"L-640471\", \"C48H74O14\", \"IV\", \"IVM-654\", \"IVR-25\", \"IV-104\", \"IVE-11\", \"IVER-15\"],\n",
    "    \"GABA\" : [\"GABA\", \"GABAergic\", \"gamma-aminobutyric acid\"],\n",
    "    \"Zebrafish\" : [\"Zebrafish\", \"Danio rerio\"],\n",
    "    \"COVID-19\" : [\"COVID-19\", \"COVID\", \"SARS-CoV-2\"],\n",
    "    \"Glutamate\" : [\"Glutamate\", \"glutamate\",\"Glu\", \"L-(+)-glutamate\",\"L-Glu\", \"L-Glutamate\", \"L-glutamate\", \"L glutamate\", \"L glutamate\"]\n",
    "}\n",
    "print(\"Done.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d0758957-98e0-47b0-bfa7-cc356b6cab88",
   "metadata": {},
   "outputs": [],
   "source": [
    "def global_db_search(query_list, genes_list=[]):\n",
    "\n",
    "    if len(genes_list) != 0:\n",
    "        for i in query_list:\n",
    "            for j in genes_list:\n",
    "                query = i + j + \"[tiab]\"\n",
    "\n",
    "                handle = Entrez.egquery(term=query)\n",
    "                record = Entrez.read(handle)\n",
    "                df = pd.DataFrame(record[\"eGQueryResult\"]).head(2)\n",
    "                df[\"Query\"] = query\n",
    "                append_data(df, 'global_query_res.csv', False)\n",
    "                time.sleep(0.34)\n",
    "    else:\n",
    "        for i in query_list:\n",
    "            handle = Entrez.egquery(term=i)\n",
    "            record = Entrez.read(handle)\n",
    "            df = pd.DataFrame(record[\"eGQueryResult\"]).head(2)\n",
    "            df[\"Query\"] = i\n",
    "            append_data(df, 'global_query_res.csv', False)\n",
    "            time.sleep(0.34)\n",
    "    return\n",
    "\n",
    "\n",
    "## Function for reading in the df \"summary\" results\n",
    "def read_in_results(file_name):\n",
    "\n",
    "    # The converters are there so that each list is NOT inside a string\n",
    "    res_df = pd.read_csv(file_name,  converters={\"MainID_List\": lambda x: x.strip(\"[]\").replace(\"'\", \"\").split(\", \"),\n",
    "                                                    \"P_Dates\": lambda x: x.strip(\"[]\").replace(\"'\", \"\").split(\", \"),\n",
    "                                                    \"P_Years\": lambda x: x.strip(\"[]\").replace(\"'\", \"\").split(\", \"),\n",
    "                                                    \"LinkedID_List\": lambda x: x.strip(\"[]\").replace(\"'\", \"\").split(\", \"),\n",
    "                                                    \"Query_Count\": int})\n",
    "    return res_df\n",
    "\n",
    "\n",
    "def esummary_info(in_webenv_key, in_query_key, db_name):\n",
    "\n",
    "    # Obtaining DocSums for a set of IDs that are stored on the Entrez History server.\n",
    "    handle = Entrez.esummary(db=db_name, webenv=in_webenv_key, query_key=in_query_key)\n",
    "    record = Entrez.read(handle)\n",
    "\n",
    "    publ_dates, publ_years = get_published_dates(record)\n",
    "\n",
    "    if db_name == \"pubmed\":\n",
    "        ids_list = get_pmcids(record)\n",
    "    else:\n",
    "        ids_list = get_pmids(record)\n",
    "\n",
    "    return publ_dates, publ_years, ids_list\n",
    "\n",
    "\n",
    "def get_published_dates(esummary_rec):\n",
    "\n",
    "    retr_dates = []\n",
    "    retr_years = []\n",
    "    check = True\n",
    "    i = 0\n",
    "    for article in esummary_rec:\n",
    "        # \"PubDate\" is often of the form: '2021 Nov 26'\n",
    "        retr_dates.append(article[\"PubDate\"])\n",
    "        date = article[\"PubDate\"].split()\n",
    "        while(i < 2):\n",
    "            if len(date[i]) == 4:\n",
    "                p_year = int(date[i])\n",
    "                i = 3\n",
    "            else:\n",
    "                i = i + 1\n",
    "        #p_year = int(article[\"PubDate\"].split()[0])\n",
    "        #p_year = int(article[\"P_Years\"].split()[0])\n",
    "        #changed from PubDate to P_Years because of error when changing search term\n",
    "        retr_years.append(p_year)\n",
    "\n",
    "    return retr_dates, retr_years\n",
    "\n",
    "\n",
    "def get_pmcids(esummary_rec):\n",
    "\n",
    "    pmcids_list = []\n",
    "    for i in esummary_rec:\n",
    "        # If \"pmc\" is there, then this article also has a PMCID (i.e., it's also found in the PubMed Central db)\n",
    "        if \"pmc\" in i[\"ArticleIds\"]:\n",
    "            pmcids_list.append(i[\"ArticleIds\"][\"pmc\"])\n",
    "        else:\n",
    "            pmcids_list.append(np.NaN)\n",
    "\n",
    "    return pmcids_list\n",
    "\n",
    "\n",
    "def get_pmids(esummary_rec):\n",
    "\n",
    "    pmids_list = []\n",
    "    for i in esummary_rec:\n",
    "        # '0' means that the article has no PMID (i.e., it's not found in the PubMed db)\n",
    "        if i[\"ArticleIds\"][\"pmid\"] == '0':\n",
    "            pmids_list.append(np.NaN)\n",
    "        else:\n",
    "            pmids_list.append(i[\"ArticleIds\"][\"pmid\"])\n",
    "\n",
    "    return pmids_list\n",
    "\n",
    "\n",
    "## Function that retrieves summary results from a given set of queries (which don't require a gene list)\n",
    "def get_query_info_no_genes(query_in, db_name):\n",
    "\n",
    "    # relevance: Records are sorted based on relevance to your search. (Relevance ranking)\n",
    "    search_results = Entrez.read(\n",
    "        Entrez.esearch(db=db_name, term=query_in, sort=\"relevance\", retmax=5000, usehistory=\"y\")\n",
    "        )\n",
    "\n",
    "    # NEED TO FIRST CHECK IF WE GOT ANY RESULTS FROM THAT QUERY\n",
    "    if len(search_results[\"IdList\"]) == 0:\n",
    "        print(\"No Results.\")\n",
    "        return\n",
    "    else:\n",
    "        # With search_results, we will use its WebEnv value and QueryKey value\n",
    "        p_dates, p_years, ids_list = esummary_info(search_results[\"WebEnv\"], search_results[\"QueryKey\"], db_name)\n",
    "\n",
    "        time.sleep(0.34)\n",
    "\n",
    "        return pd.DataFrame([[query_in, db_name, search_results['Count'], search_results['IdList'], p_dates, p_years, ids_list]],\n",
    "                                columns=['Query', 'Db_Name', 'Query_Count', 'MainID_List', 'P_Dates', 'P_Years', 'LinkedID_List'])\n",
    "\n",
    "\n",
    "## Function that retrieves summary results from a given set of queries (which requires a gene list)\n",
    "def get_query_info(query_in, genes, db_name):\n",
    "\n",
    "    gene_query = []\n",
    "    query = \"\"\n",
    "\n",
    "    for i in genes:\n",
    "        # Example of db_name values in this use case: \"pubmed\" or \"pmc\"\n",
    "        if db_name == \"pubmed\":\n",
    "            # PubMed's Search field tag: Title/Abstract [tiab]\n",
    "            query = query_in + i + \"[tiab]\"\n",
    "        else:\n",
    "            query = query_in + i\n",
    "\n",
    "        # relevance: Records are sorted based on relevance to your search. (Relevance ranking)\n",
    "        search_results = Entrez.read(\n",
    "            Entrez.esearch(db=db_name, term=query, sort=\"relevance\", retmax=5000, usehistory=\"y\")\n",
    "            )\n",
    "     \n",
    "        # NEED TO FIRST CHECK IF WE GOT ANY RESULTS FROM THAT QUERY\n",
    "        if len(search_results[\"IdList\"]) == 0:\n",
    "            continue\n",
    "\n",
    "        # With search_results, we will use its WebEnv value and QueryKey value\n",
    "        p_dates, p_years, ids_list = esummary_info(search_results[\"WebEnv\"], search_results[\"QueryKey\"], db_name)\n",
    "       \n",
    "        gene_query.append([query, db_name, search_results['Count'], search_results['IdList'], p_dates, p_years, ids_list])\n",
    "        time.sleep(0.34)\n",
    "          \n",
    "    return pd.DataFrame(gene_query, columns=['Query', 'Db_Name', 'Query_Count', 'MainID_List', 'P_Dates', 'P_Years', 'LinkedID_List'])\n",
    "\n",
    "\n",
    "## Function for obtaining citation counts for the set of IDs found in the \"summary\" df\n",
    "def cited_cnt_table(df_summary, db_name):\n",
    "\n",
    "    elink_data = []\n",
    "    link_name = \"\"\n",
    "\n",
    "    if db_name == \"pubmed\":\n",
    "        link_name = \"pubmed_pubmed_citedin\"\n",
    "    else:\n",
    "        link_name = \"pmc_pmc_citedby\"  # \"pmc\" is the other db_name in this use case\n",
    "\n",
    "    for i in range(0, len(df_summary)):\n",
    "\n",
    "        query_term = df_summary.iloc[i][\"Query\"]\n",
    "\n",
    "        for id_num in df_summary.iloc[i][\"MainID_List\"]:\n",
    "\n",
    "            record = Entrez.read(Entrez.elink(id=id_num, dbfrom=db_name, db=db_name, linkname=link_name))\n",
    "         \n",
    "            if len(record[0][\"LinkSetDb\"]) != 0:\n",
    "                cited_counts = len(record[0][\"LinkSetDb\"][0][\"Link\"])\n",
    "            else:\n",
    "                # 'LinkSetDb' key contains empty list when an article has no citation counts\n",
    "                cited_counts = 0\n",
    "            elink_data.append([query_term, db_name, id_num, cited_counts])\n",
    "\n",
    "            if (df_summary.iloc[i][\"MainID_List\"].index(id_num) + 1) % 3 == 0:\n",
    "                time.sleep(0.34)\n",
    "\n",
    "    return pd.DataFrame(elink_data, columns=[\"Query\", \"Db_Name\", \"Id_List\", \"Citation_Cnts\"]) \n",
    "\n",
    "\n",
    "## Function that returns the Top-k results (pass in k as an argument to the function, input by the user)\n",
    "def get_top_k(df, k_val):\n",
    "\n",
    "    q_top_k = []\n",
    "\n",
    "    for q in df[\"Query\"].unique():\n",
    "        matches_ids = []  # For each query version, these are the IDs meeting the criteria of having citation counts >= 25\n",
    "        counts = []\n",
    "        df_temp = df[df[\"Query\"] == q]\n",
    "\n",
    "        for i in range(0, len(df_temp)):\n",
    "            if df_temp.iloc[i][\"Citation_Cnts\"] >= 25:\n",
    "                matches_ids.append(int(df_temp.iloc[i][\"Id_List\"]))\n",
    "                counts.append(df_temp.iloc[i][\"Citation_Cnts\"])\n",
    "                if len(matches_ids) == k_val:\n",
    "                    break\n",
    "        if len(matches_ids) == 0:\n",
    "            continue\n",
    "        q_top_k.append([q, matches_ids, counts])\n",
    "\n",
    "    top_k_df = pd.DataFrame(q_top_k, columns=[\"Query\", \"Top_\"+str(k_val)+\"_Ids\", \"Citation_Cnts\"])\n",
    "\n",
    "    return top_k_df\n",
    "\n",
    "\n",
    "## Function that appends DataFrame rows to a CSV file\n",
    "def append_data(df, file_name, is_new_file):\n",
    "\n",
    "    if is_new_file:\n",
    "        # if True, then\n",
    "        df.to_csv(file_name, index=False)\n",
    "    else:\n",
    "        # False: This is an existing CSV file\n",
    "        df.to_csv(file_name, mode='a', index=False, header=False)\n",
    "\n",
    "    return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b15ad664-00ed-45c2-b82d-6bdb63797f7c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done.\n"
     ]
    }
   ],
   "source": [
    "# Load in the model for English\n",
    "nlp = spacy.load(\"en_core_web_sm\")\n",
    "# Can't retrieve XML if you don't have a query.\n",
    "query = \"GABA AND Glutamate\"\n",
    "Entrez.email = \"n01365801@unf.edu.com\"\n",
    "df_q_pubmed = get_query_info_no_genes(query, \"pubmed\")\n",
    "print(\"Done.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "8d0a4fb4-3e76-49d6-8137-e30c4b853bf1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done.\n"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "Documentation for fetch_pubmed()\n",
    "\n",
    "This function's purpose is to use NCBI's E-Utils to get the body of articles, given an id.\n",
    "The E-Util used in E-Fetch.\n",
    "Future Work on this could be adjusting the argument, to allow for just a list of IDs, instead of a Pandas DataFrame slice.\n",
    "\n",
    "Arguments:\n",
    "    * ids: The IDs of articles from a Pandas DataFrame\n",
    "Return Value: A list of the records.\n",
    "\"\"\"\n",
    "def fetch_pubmed(ids): \n",
    "    records_pubmed = []\n",
    "    # Fetch all records pertaining to our queries.\n",
    "    for row in ids:\n",
    "        for uid in row:\n",
    "            handle = Entrez.efetch(db=\"pubmed\", id=uid, retmode=\"xml\")\n",
    "            record = Entrez.read(handle, validate=False)\n",
    "            records_pubmed.append(record)\n",
    "    # Be polite and flush/close the stream like a good programmer.\n",
    "    handle.close()\n",
    "    return records_pubmed\n",
    "print(\"Done.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "6f2a3a7c-60ce-4871-8fae-4aa2651ea715",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done.\n"
     ]
    }
   ],
   "source": [
    "records_pubmed = fetch_pubmed(df_q_pubmed['MainID_List'])\n",
    "print(\"Done.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "085aaa40-18e0-421b-9b36-cd3762e128b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "#print(records_pubmed[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "bfcd7d35-d972-4502-8db5-6cbc884061da",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done.\n"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "Documentation for invert_dict()\n",
    "\n",
    "This function's sole purpose is to invert a dictionary, so that the values of the old are the keys of the new,\n",
    "and the keys of the old are the values of the new, in list format.\n",
    "\n",
    "Arguments:\n",
    "    dictionary: The dictionary to be inverted.\n",
    "Return Value: The inverted dictionary following the above design.\n",
    "\"\"\"\n",
    "def invert_dict(dictionary):\n",
    "    dict_inverted = {} # output\n",
    "    for (k, v) in dictionary.items():\n",
    "        if v in dict_inverted.keys():\n",
    "            dict_inverted[v].append(k)\n",
    "        else:\n",
    "            dict_inverted[v] = [k]\n",
    "            \n",
    "    return dict_inverted\n",
    "print(\"Done.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "529df7e7-04f5-4f14-8c7e-d9162e97c46e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done.\n"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "Documentation for find_comentions()\n",
    "\n",
    "Future Work for this function includes generalizing it to be able to handle both PubMed and PMC.\n",
    "This will likely require some work on reranking() below, as it only handles PubMed formatted XML,\n",
    "due to issues with PMC and Biopython present while writing this code.\n",
    "\n",
    "Arguments:\n",
    "    * thes: A dictionary containing the synonyms of query terms.\n",
    "    * doc: A SpaCy Doc object that contains the text we are looking at.\n",
    "Return Value: A tuple in the form of (sentences, proximity list)\n",
    "\"\"\"\n",
    "def find_comentions(thes, doc):\n",
    "    sentences = []\n",
    "    proximity_list = []\n",
    "    for sentence in doc.sents:\n",
    "        prev_term = \"\"\n",
    "        term_seen = False\n",
    "        first_i = 0\n",
    "        for word in sentence:\n",
    "            for term in thes.keys():\n",
    "                if term_seen:\n",
    "                    if (word.text in thes[term]) and (word.text not in thes[prev_term]):\n",
    "                        proximity_list.append(int(word.i - first_i))\n",
    "                        first_i = word.i\n",
    "                        prev_term = term\n",
    "                        sentences.append(str(sentence))\n",
    "                        break\n",
    "                elif (word.text in thes[term]):\n",
    "                    term_seen = True\n",
    "                    first_i = word.i\n",
    "                    prev_term = term\n",
    "                    break\n",
    "    \n",
    "    return (str(sentences), proximity_list)\n",
    "print(\"Done.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "27658a40-2d25-470b-9e63-befb4fab9b9a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Record indexed at 700 improperly formatted/index error.\n",
      "Record indexed at 3600 improperly formatted/index error.\n",
      "Record indexed at 4836 improperly formatted/index error.\n",
      "Num records proccessed: 5000\n"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "Documentation for reranking()\n",
    "\n",
    "This function will use find_comentions(), invert_dict(), and word_proximity() to create a general ranking of the articles.\n",
    "The articles will be identified by UID.\n",
    "\n",
    "Future additions to this reranking function include using Ms. Victoria's get_top_k() and cited_cnt_table() functions to add in the\n",
    "25 citation requirement for credibility. Other work includes breaking out some functionality into other functions in order to clean up the mess.\n",
    "\n",
    "Arguments:\n",
    "    * records: A list of XML Objects returned by E-Fetch.\n",
    "    * query_terms: A list of query_terms. These then get selected out of the Thesaurus.\n",
    "Return Value: The rankings for the records. Type is a Pandas DataFrame.\n",
    "\"\"\"\n",
    "### TODO: Break apart this function into smaller functions\n",
    "    # Namely, creation of combined_criteria, creation of top_15, and creation of DataFrame\n",
    "def reranking(records, query_terms):\n",
    "    # snippet from https://stackoverflow.com/questions/29216889/slicing-a-dictionary\n",
    "    # While in the answer, they add a check to make sure the key is in original dict\n",
    "    # (the thesaurus in this case), it is safe to assume that the key is in the dict,\n",
    "    # because the thesaurus should contain all possible terms.\n",
    "    inner_thesaurus = {k:thesaurus[k] for k in query_terms}\n",
    "\n",
    "    # DataFrame Data\n",
    "    pmids = []\n",
    "    titles = []\n",
    "    abstracts = []\n",
    "    relevancy_score = {}\n",
    "    comention_sentences = []\n",
    "    i = 0\n",
    "    j = 0\n",
    "    for record in records:\n",
    "        # Some records do not have an Abstract (???) so we need to check for an abstract\n",
    "        # TODO: Deal with the articles that don't have abstracts\n",
    "        try:\n",
    "            pmid = str(record['PubmedArticle'][0]['MedlineCitation']['PMID'])\n",
    "            article_keys = record['PubmedArticle'][0]['MedlineCitation']['Article'].keys()\n",
    "            if 'Abstract' in article_keys: # We have an Abstract\n",
    "                # TODO: Stop rewriting this indexing mess every time.\n",
    "                abstract_text = str(record['PubmedArticle'][0]['MedlineCitation']['Article']['Abstract']['AbstractText'])\n",
    "                # Process the abstract\n",
    "                doc = nlp(abstract_text)\n",
    "\n",
    "                (comention_sents, proximity_count) = find_comentions(doc=doc, thes=inner_thesaurus)\n",
    "\n",
    "                # DataFrame data collection\n",
    "                pmids.append(pmid)\n",
    "                if (len(proximity_count) != 0):\n",
    "                    relevancy_score[pmid] = sum([1/count for count in proximity_count])\n",
    "                else:\n",
    "                    relevancy_score[pmid] = 0\n",
    "                titles.append(str(record['PubmedArticle'][0]['MedlineCitation']['Article']['ArticleTitle']))\n",
    "                try:\n",
    "                    #FIXME: have to repeat these commands in order to have proper formatting \n",
    "                    abstract_text = \"\".join(str(i) for i in (record['PubmedArticle'][0]['MedlineCitation']['Article']['Abstract']['AbstractText']))\n",
    "                    abstracts.append(abstract_text)\n",
    "                    doc = nlp(abstract_text)\n",
    "                    (comention_sents, proximity_count) = find_comentions(doc=doc, thes=inner_thesaurus)\n",
    "                    comention_sentences.append(comention_sents)\n",
    "                except:\n",
    "                    print(\"Reformatting error at \" + str(i) + \".\")\n",
    "                \n",
    "        except:\n",
    "            print(\"Record indexed at \" + str(i) + \" PMID not found.\")\n",
    "        i = i + 1\n",
    "        j = j + 1\n",
    "    print(\"Num records proccessed: \" + str(j))\n",
    "    \n",
    "    # Relevancy score will be used to calculate the index.\n",
    "    prerankings = pd.DataFrame(data=[pmids, titles, abstracts, relevancy_score.values(), comention_sentences]).transpose()\n",
    "    prerankings.columns = [\"PMID\", \"Title\", \"Abstract\", \"Relevancy Score\", \"Comention Sentences\"]\n",
    "    \n",
    "    \n",
    "    #formatting for output\n",
    "    \n",
    "    \n",
    "    relevancy_score_inverted = invert_dict(relevancy_score)\n",
    "    # Gathering the top 15 abstracts based on the relevancy score.\n",
    "    # We conglomerate all of the scores into a single list, flatten it, and then simply slice out the first 15 elements.\n",
    "    top_15 = []\n",
    "    for i in sorted(relevancy_score_inverted.keys(), reverse=True):\n",
    "        top_15.append(relevancy_score_inverted[i])\n",
    "    # This list comprehension is flattening the list of lists produced by combined_criteria filtering.\n",
    "    top_15 = [x for xs in top_15 for x in xs]\n",
    "    top_15 = top_15[:15]\n",
    "    prerankings.astype({'Comention Sentences': 'string'}).dtypes\n",
    "    prerankings[\"Comention Sentences\"] = prerankings[\"Comention Sentences\"].apply(lambda x: \"\".join(x))\n",
    "    prerankings[\"Comention Sentences\"] = prerankings[\"Comention Sentences\"].apply(lambda x: str(x).strip(\"[]\"))\n",
    "    #prerankings[\"Title\"] = prerankings[\"Title\"].apply(lambda x: str(x).strip(\"[]\"))\n",
    "    \n",
    "    \n",
    "    # Now that we have the UIDs of our top 15, we can grab them\n",
    "    rankings = prerankings[prerankings[\"PMID\"].isin(top_15)].sort_values(by=\"Relevancy Score\", ascending=False).reset_index(drop=True)\n",
    "    \n",
    "    return rankings\n",
    "\n",
    "# FIXME: Do NOT hardcode the query_terms, pull them in from the query.\n",
    "rankings = reranking(records_pubmed, [\"GABA\",\"Glutamate\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "620907f5-4fb2-4a5b-a1e5-29913147c7a1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done.\n"
     ]
    }
   ],
   "source": [
    "rankings\n",
    "print(\"Done.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "e9ed846c-de2f-439e-a6d5-dec2fcd16319",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done.\n"
     ]
    }
   ],
   "source": [
    "# CSV output\n",
    "rankings.to_csv(\"Glutamate and GABA5.csv\")\n",
    "print(\"Done.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "6558c7e9-f2fe-411c-bf2e-83b33e7b5ae3",
   "metadata": {},
   "outputs": [],
   "source": [
    "REST_URL = \"http://data.bioontology.org\"\n",
    "\"\"\"\n",
    "Use of this requires a valid API Key.\n",
    "This simply requires setting up an account at bioportal.bioonotology.org,\n",
    "but do note that API Keys are meant to be secret, and probably more permanent than\n",
    "what amounts to basically an intern's key.\n",
    "\"\"\"\n",
    "API_KEY = \"0de90ddd-0da2-42ee-b0a0-2eb3b90dface\"\n",
    "ont = \"HP,MDM\"\n",
    "def get_json(url):\n",
    "    opener = urllib.request.build_opener()\n",
    "    opener.addheaders = [('Authorization', 'apikey token=' + API_KEY)]\n",
    "    return json.loads(opener.open(url).read())\n",
    "\n",
    "abstract = \"Ivermectin (IVM) is an antiparasitic drug that is widely used in domestic animals. In mammals, IVM acts as a γ-aminobutyric acid (GABA) receptor agonist. This neurotransmitter plays an important role in the regulation of female sexual behavior. The present study investigated the effects of therapeutic (0.2 mg/kg) and high (1.0 mg/kg) IVM doses on female sexual behavior in physiological and pharmacological conditions. Female rats in estrus or treated with estradiol valerate to induce sexual behavior 24 h before the experiments were used. Ivermectin was administered 15 min before the sexual observations. The number of lordosis events in 10 mounts was recorded to calculate the lordosis quotient. The intensity of lordosis (0 [no lordosis], 1 [low lordosis], 2 [normal lordosis] and 3 [exaggerated lordosis]) was scored. In estrus and hormonal treated female rats, both IVM doses decreased the intensity of the lordosis reflex and the percentage of females that presented high levels of lordosis (exaggerated lordosis). However, the number of females that presented lordosis was unaltered. We conclude that in both hormonal conditions, 0.2mg/kg IVM treatment reduced female sexual behavior and the execution of the lordosis reflex. The present results may be useful for avoiding the side effects of this drug in veterinary practice.\"\n",
    "annotations = get_json(REST_URL + \"/annotator?text=\" + urllib.parse.quote(abstract) + \"&ontologies=\" + ont)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "8a8b451f-15d1-488a-82f0-b2ecc90c3428",
   "metadata": {},
   "outputs": [],
   "source": [
    "def print_annotations(annotations, get_class=True):\n",
    "    for result in annotations:\n",
    "        class_details = result[\"annotatedClass\"]\n",
    "        if get_class:\n",
    "            try:\n",
    "                class_details = get_json(result[\"annotatedClass\"][\"links\"][\"self\"])\n",
    "            except urllib.error.HTTPError:\n",
    "                print(f\"Error retrieving {result['annotatedClass']['@id']}\")\n",
    "                continue\n",
    "        print(\"Class details\")\n",
    "        print(\"\\tid: \" + class_details[\"@id\"])\n",
    "        print(\"\\tprefLabel: \" + class_details[\"prefLabel\"])\n",
    "        print(\"\\tontology: \" + class_details[\"links\"][\"ontology\"])\n",
    "\n",
    "        print(\"Annotation details\")\n",
    "        for annotation in result[\"annotations\"]:\n",
    "            print(\"\\tfrom: \" + str(annotation[\"from\"]))\n",
    "            print(\"\\tto: \" + str(annotation[\"to\"]))\n",
    "            print(\"\\tmatch type: \" + annotation[\"matchType\"])\n",
    "\n",
    "        if result[\"hierarchy\"]:\n",
    "            print(\"\\n\\tHierarchy annotations\")\n",
    "            for annotation in result[\"hierarchy\"]:\n",
    "                try:\n",
    "                    class_details = get_json(annotation[\"annotatedClass\"][\"links\"][\"self\"])\n",
    "                except urllib.error.HTTPError:\n",
    "                    print(f\"Error retrieving {annotation['annotatedClass']['@id']}\")\n",
    "                    continue\n",
    "                pref_label = class_details[\"prefLabel\"] or \"no label\"\n",
    "                print(\"\\t\\tClass details\")\n",
    "                print(\"\\t\\t\\tid: \" + class_details[\"@id\"])\n",
    "                print(\"\\t\\t\\tprefLabel: \" + class_details[\"prefLabel\"])\n",
    "                print(\"\\t\\t\\tontology: \" + class_details[\"links\"][\"ontology\"])\n",
    "                print(\"\\t\\t\\tdistance from originally annotated class: \" + str(annotation[\"distance\"]))\n",
    "\n",
    "        print(\"\\n\\n\")\n",
    "        \n",
    "#print_annotations(annotations)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "c9422760-b048-4e5a-bb37-b4cadbe3a072",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"def get_json(url):\\n    opener = urllib.request.build_opener()\\n    opener.addheaders = [('Authorization', 'apikey token=' + API_KEY)]\\n    return json.loads(opener.open(url).read())\\n\""
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import inspect\n",
    "inspect.getsource(get_json)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "id": "613d9952-582d-4ce3-ac3a-5293c56bf349",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ABSTRACT 1\n",
      "ABSTRACT 2\n",
      "ABSTRACT 3\n",
      "ABSTRACT 4\n",
      "ABSTRACT 5\n",
      "ABSTRACT 6\n",
      "ABSTRACT 7\n",
      "ABSTRACT 8\n",
      "ABSTRACT 9\n",
      "ABSTRACT 10\n",
      "ABSTRACT 11\n",
      "ABSTRACT 12\n",
      "ABSTRACT 13\n",
      "ABSTRACT 14\n",
      "ABSTRACT 15\n",
      "[{'HP': ['Central', 'Schizophrenia', 'Clinical course', 'Peripheral'], 'MDM': ['gamma-Aminobutyric Acid, Monolithium Salt', 'gamma-Aminobutyric acid', 'D-Glutamate', 'Neurohormones', 'Central Nervous System', 'Nervous System', 'Anxiousness', 'Clients', 'Schizophrenia', 'Fresh Frozen Plasma', 'Surrogate Markers', 'Heredity', 'Risperidone', 'R-64,766', 'Antipsychotic Effect', 'Therapeutics', 'Normality', 'R 76477', 'Paliperidone', 'Syndrome', 'Cluster Analysis', 'Laboratory Research', 'Clinical Investigators']}, {'HP': ['Lateral', 'Depression', 'Schizophrenia'], 'MDM': ['Puerperium', 'Brain', 'Compassion', 'Growth and Development', 'Growth', 'D-Glutamate', 'gamma-Aminobutyric Acid, Monolithium Salt', 'gamma-Aminobutyric acid', 'Scales', 'Neurohormones', 'Phosphate-Activated Glutaminase', 'D-Glutamine', 'D-Glutamine', 'L-Glutamine', 'Female', 'Mice, Laboratory', 'Tissues', 'mRNA, Polyadenylated', 'Enzymes', 'Glutamate Decarboxylase', 'Carboxy-Lyases', 'Glutamate-Ammonia Ligase', 'Ammonia', 'Ammonia', 'Ligases', 'Postpartum Abstinence', 'Theses', 'Readability', 'Mental Health', 'Mentalization', 'Normality', 'Depressive Symptoms', 'Schizophrenia', 'Mental Disorders, Severe', \"Mothers' Clubs\"]}, {'HP': [], 'MDM': ['Ventral Tegmental Area', 'Neurons', 'Playthings', 'Role Concept', 'Reward', 'D-Glutamate', 'gamma-Aminobutyric Acid, Monolithium Salt', 'gamma-Aminobutyric acid', 'Theses', 'Cells', 'Behaviorism', 'Acceptance Processes', 'Resolvases', 'Mice, Laboratory', 'Cues', 'GABAergic Neurons']}, {'HP': ['Affected'], 'MDM': ['Propofol Abbott', 'Propofol', 'D-Glutamate', 'gamma-Aminobutyric Acid, Monolithium Salt', 'gamma-Aminobutyric acid', 'Aminobutyric Acids', 'Acids', 'Synaptosomes', 'Subiculum', 'Cerebrospinal Fluid', 'Observation', 'Veratridine', 'Chromatography, High Pressure Liquid', 'Chromatography, Liquid', 'Chromatography']}, {'HP': ['Acute', 'Healthy'], 'MDM': ['gamma-Aminobutyric Acid, Monolithium Salt', 'gamma-Aminobutyric acid', 'D-Glutamate', 'Brain', 'Lamina Terminalis', 'Mood', 'Castor Beaver', 'Humans', 'Humanities', 'Humanism', 'Trier Social Stress Test', 'Socialization', 'Socialism', 'Sociality', 'Bicycle Ergometry Test', 'Testis', 'Anterior Prefrontal Cortex', 'Male', 'Malingering', '1H NMR', 'Protons', 'Spectroscopy, Nuclear Magnetic Resonance', 'Magnetics', 'Spectroscopy', 'Association', 'Hydrocortisone', 'Cortril', 'Longterm Effects', 'Overall']}, {'HP': ['Diffuse'], 'MDM': ['Brain', 'Microdialysis', 'Procedures', 'Neurohormones', 'Dialysis Solutions', 'Abstracting', 'Index', 'Theses', 'Dopamine', 'Intropin', 'Hydroxytryptamine', 'Serotonin', 'Acetylcholine Sulfate (1:1)', 'Acetylcholine', 'D-Glutamate', 'gamma-Aminobutyric Acid, Monolithium Salt', 'gamma-Aminobutyric acid', 'Originality', 'Extracellular Space', 'Heredity', 'Therapeutics', 'Astrocytes', 'Receptors, GABA', 'Synapses', 'Conception']}, {'HP': ['Psychosis', 'Healthy'], 'MDM': ['Brain', 'D-Glutamate', 'gamma-Aminobutyric Acid, Monolithium Salt', 'gamma-Aminobutyric acid', 'Aminobutyric Acids', 'Acids', 'Homeostasis', 'Psychosis, Brief Reactive', 'Relative Risk', 'Epistemology', 'Neurochemistry', 'Meta-Analysis', 'Index Medicus', 'Phosphate-Activated Glutaminase', 'D-Glutamine', 'D-Glutamine', 'L-Glutamine', 'Sample Size', 'Frontal Eye Fields', 'Family Research', 'Health Services Needs']}, {'HP': [], 'MDM': ['Neurons', 'Physically Challenged', 'Neurohormones', 'D-Glutamate', 'gamma-Aminobutyric Acid, Monolithium Salt', 'gamma-Aminobutyric acid', 'Aminobutyric Acids', 'Acids', 'Levoglucose', 'D-glucose', 'Glucose, (beta-D)-Isomer', 'Dextrose, unspecified form', 'Phosphate-Activated Glutaminase', 'D-Glutamine', 'D-Glutamine', 'L-Glutamine', 'Astrocytes', 'Review Literature', 'Commuting', 'Catabolism', 'Role Concept', 'Ammonia', 'Ammonia', 'Homeostasis', 'Maine', 'Vitreous Carbon', 'Carbonates', 'Norit', 'Activated charcoal', 'Carbonate ion', 'Nitrogen', 'Nitrogen']}, {'HP': ['Bipolar affective disorder', 'Mania', 'Depression', 'Postural instability', 'Healthy'], 'MDM': ['Manic Disorder', 'Mood', 'Manic Episode', 'Depressive Symptoms', 'gamma-Aminobutyric Acid, Monolithium Salt', 'gamma-Aminobutyric acid', 'Synaptic Transmission', '1H NMR', 'Protons', 'Spectroscopy, Nuclear Magnetic Resonance', 'Magnetics', 'Spectroscopy', 'Theses', 'Enzymes', 'Glutamate Decarboxylase', 'Glutamic acid', 'Glutamic Acids', 'D-Glutamate', 'Acids', 'Carboxy-Lyases', 'Decarboxylation', 'Polymorphism, Single Nucleotide', 'Nucleotides', 'Brodmann Area 24', 'Clients', 'Genetic Materials', 'Antiepileptic Agents', 'Antipsychotic Effect', 'Lithium', 'Alleles', 'Role Concept', 'Genogroup']}, {'HP': ['Healthy', 'Stable'], 'MDM': ['Spectroscopy, Nuclear Magnetic Resonance', 'Magnetics', 'Spectroscopy', 'Scales', 'Neurohormones', 'gamma-Aminobutyric Acid, Monolithium Salt', 'gamma-Aminobutyric acid', 'D-Glutamate', 'Humans', 'Humanities', 'Humanism', 'Brain', 'Role Concept', 'Behaviorism', 'Acceptance Processes', 'Extrastriate Cortex', 'Sens-A-Ray', 'Eye', 'Darkness', 'Photic Stimulation', 'Spin Echo Imaging', 'Neurofibromin 2', 'Isoxaflutole', 'Playthings', 'Dissent', 'Work', 'Visual Perception']}, {'HP': ['Depression', 'Moderate', 'Behavioral abnormality'], 'MDM': ['Clients', 'Pathology', 'Depressive Symptoms', 'Frontal Eye Fields', 'gamma-Aminobutyric Acid, Monolithium Salt', 'gamma-Aminobutyric acid', 'D-Glutamate', 'Phenotype', 'Anterior Prefrontal Cortex', 'Behaviorism', 'Acceptance Processes', 'Rats, Laboratory', 'Testis', 'Magic', 'Spectroscopy, Nuclear Magnetic Resonance', 'Magnetics', 'Spectroscopy', 'mRNA, Polyadenylated', 'Fresh Frozen Plasma', 'Corticosterone', 'Corticosterone', 'SGPT', 'Abufène', 'Alanine', 'Transaminases', 'SGOT', 'Potassium aspartate', 'Aspartic acid', 'Magnesium aspartate', 'Dipotassium Aspartate', 'Macro-Creatine Kinase', 'Creatine', 'Creatine', 'Phosphotransferases, ATP', 'L-Lactate Dehydrogenase', 'L-Lactic acid', 'Lactic acid', 'L-Lactic Acid', 'Ammonium lactate', 'D-Lactic acid', 'Oxidase', 'Locomotion', 'Theses', 'TOR Serine-Threonine Kinases', 'Rapamune', 'Sirolimus', 'Adenine Nucleotides', 'Adenosine phosphate', 'Phosphaden', 'Adenoscan', 'Deoxyadenosines', 'Adenosine', 'Protein Kinases', 'Gene Products, Protein', 'Metazoa', 'Individuality']}, {'HP': [], 'MDM': ['Methylation', 'Pyrrolidines', 'gamma-Aminobutyric Acid, Monolithium Salt', 'gamma-Aminobutyric acid', 'Commuting', 'Membranes', 'Inhibitory Concentration 50', 'Theses', 'Esters', 'GABA Transporter 2']}, {'HP': ['Involuntary movements', 'Transient ischemic attack', 'Postural instability'], 'MDM': ['Syndrome', 'Ballismus', 'Movement', 'Neurotic Disorders', 'D-Glutamate', 'gamma-Aminobutyric Acid, Monolithium Salt', 'gamma-Aminobutyric acid', 'Aminobutyric Acids', 'Acids', 'Phosphate-Activated Glutaminase', 'D-Glutamine', 'D-Glutamine', 'L-Glutamine', 'Regulation', 'FLO-1347', 'Tiapride', 'Male', 'Malingering', 'Rats, Wistar', 'Rats, Laboratory', 'Control Groups', 'Amino Acids', 'Amino acids', 'Neurohormones', 'Potassium aspartate', 'Aspartic acid', 'Magnesium aspartate', 'Dipotassium Aspartate', 'Glycine, Monoammonium Salt', 'Glycine', 'Glutamate-Ammonia Ligase', 'Ligases', 'Glutamate Transporter', 'Glutamate Decarboxylase', 'Carboxy-Lyases', 'Gene Products, Protein', 'Blot, Western', 'Quantitative Real-Time Polymerase Chain Reaction', 'Longterm Effects', 'Anchored PCR', 'Behavior Test', 'Behaviorism', 'Acceptance Processes', 'Testis', 'mRNA, Polyadenylated']}, {'HP': ['Mild'], 'MDM': ['Experimental Animal Models', 'Metazoa', 'Concussion, Mild', 'Brain Injuries, Traumatic', 'Brain Injuries, Focal', 'Brain', 'Wounds', 'Suggestion', 'Head', '1H NMR', 'Protons', 'Spectroscopy, Nuclear Magnetic Resonance', 'Magnetics', 'Spectroscopy', 'Humans', 'Humanities', 'Humanism', 'D-Glutamate', 'gamma-Aminobutyric Acid, Monolithium Salt', 'gamma-Aminobutyric acid', 'Primary Motor Cortex', 'Brodmann Area 9', 'Anterior Prefrontal Cortex', 'Longterm Effects', 'Theses', 'Epistemology', 'Neurohormones']}, {'HP': ['Stable'], 'MDM': ['Terfenadine', 'Hisfedin', 'Neurohormones', 'Mice, Laboratory', 'Brain', 'Data Analysis', 'Models, Space', 'Glutamic acid', 'D-Glutamate', 'Glutamic Acids', 'Acids', 'gamma-Aminobutyric Acid, Monolithium Salt', 'gamma-Aminobutyric acid', 'Aminobutyric Acids', 'Tandem Mass Spectrometry', 'Mass Spectrometry', 'Spectroscopy', 'Microdialysis', 'Standardization', 'Calibration', 'Isotopes', 'Procedures', 'Quality Control', 'Longterm Effects', 'Encephalopathy', 'Disease', 'Parkinson Disease', 'Autosomal Recessive Parkinsonism', 'Akinetic-Rigid Variant of Huntington Disease', 'Futurology']}]\n"
     ]
    }
   ],
   "source": [
    "def hpo_extraction(ont):\n",
    "    i = 0\n",
    "    rankedAnnotations = []\n",
    "    \n",
    "    for abstract in rankings[\"Abstract\"]:\n",
    "        i = i +1\n",
    "        print(\"ABSTRACT \" + str(i))\n",
    "        annotations = (get_json(REST_URL + \"/annotator?text=\" + urllib.parse.quote(abstract) + \"&ontologies=\" + ont))\n",
    "        rankedAnnotations.append(getAnnotations(annotations, ont))\n",
    "        #print(annotations)\n",
    "        #print(abstract + \"\\nNumber: \" + str(i))\n",
    "        #print_annotations(annotations)\n",
    "    #print(annotations)   \n",
    "    return rankedAnnotations\n",
    "    #return\n",
    "def getAnnotations(annotations, ont):\n",
    "    ontCat = ont.split(\",\")\n",
    "    ontDict={}\n",
    "    for ont in ontCat:\n",
    "        ontDict[ont] = []\n",
    "\n",
    "    for result in annotations:\n",
    "        class_details = result[\"annotatedClass\"]\n",
    "        try:\n",
    "            class_details = get_json(result[\"annotatedClass\"][\"links\"][\"self\"])\n",
    "        except urllib.error.HTTPError:\n",
    "            print(f\"Error retrieving {result['annotatedClass']['@id']}\")\n",
    "            continue\n",
    "\n",
    "        #print(\"Class details\")\n",
    "        #print(\"\\tid: \" + class_details[\"@id\"])\n",
    "        #print(\"\\tprefLabel: \" + class_details[\"prefLabel\"])\n",
    "        #print(\"\\tontology: \" + class_details[\"links\"][\"ontology\"])\n",
    "        #print(str(class_details[\"links\"][\"ontology\"]).split(\"/\")[-1])\n",
    "        #print(class_details[\"prefLabel\"])\n",
    "        #if ontDict[str(class_details[\"links\"][\"ontology\"].split(\"/\")[-1])]:\n",
    "        #print(type(class_details[\"prefLabel\"]))\n",
    "        ontDict[str(class_details[\"links\"][\"ontology\"].split(\"/\")[-1])] += [class_details[\"prefLabel\"]]\n",
    "        \n",
    "        #else:\n",
    "            #ontDict[str(class_details[\"links\"][\"ontology\"].split(\"/\")[-1])] = ontDict[str(class_details[\"links\"][\"ontology\"].split(\"/\")[-1])].append(str(class_details[\"prefLabel\"]))\n",
    "    return ontDict\n",
    "#hpo_extraction(ont)\n",
    "print(hpo_extraction(ont))\n",
    "#annotations = hpo_extraction()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "93072f28-6012-4240-a606-ca68004cfd4a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "γ-aminobutyrate (GABA) and Glutamate (Glu) are respectively two major inhibitory and excitatory neurotransmitters in the central nervous system and recent theories propose that both of their signaling complexes are compromised in patients with schizophrenia.The changes in plasma GABA, Glu and GABA/Glu ratio in schizophrenia have been studied and may be potential clinical markers. Here, we examined if plasma GABA, Glu and GABA/Glu ratio are altered in 32 schizophrenics, including a comprehensive investigation of their involvements with clinical course of a 6-week risperidone antipsychotic treatment.Plasma levels of GABA and Glu were significantly lower in patients than in controls, while plasma GABA/Glu ratio was significantly elevated. During treatment, a non-significant further decrease of plasma GABA, a significant increase of plasma Glu and a significant reduction of plasma GABA/Glu ratio were observed. The ratio returned to the control level at week 6 even though concentrations of GABA and Glu were still distant from normal. After the Bonferroni correction, partial correlation analyses showed that plasma GABA and GABA/Glu ratio were positively correlated with the dose of risperidone and plasma concentration of 9-hydroxyrisperidone. The reduction of plasma GABA/Glu ratio was positively correlated with the improvement of activation symptom cluster.The elevated plasma GABA/Glu ratio reinforces the idea of an abnormal GABA-Glu interaction in schizophrenia. The ratio may be a good peripheral state-like marker in schizophrenia research.\n",
      "Class details\n",
      "\tid: http://purl.obolibrary.org/obo/HP_0030645\n",
      "\tprefLabel: Central\n",
      "\tontology: https://data.bioontology.org/ontologies/HP\n",
      "Annotation details\n",
      "\tfrom: 122\n",
      "\tto: 128\n",
      "\tmatch type: PREF\n",
      "\n",
      "\n",
      "\n",
      "Class details\n",
      "\tid: http://purl.obolibrary.org/obo/MP_0008912\n",
      "\tprefLabel: nervous\n",
      "\tontology: https://data.bioontology.org/ontologies/MP\n",
      "Annotation details\n",
      "\tfrom: 130\n",
      "\tto: 136\n",
      "\tmatch type: PREF\n",
      "\n",
      "\n",
      "\n",
      "Class details\n",
      "\tid: http://purl.obolibrary.org/obo/HP_0100753\n",
      "\tprefLabel: Schizophrenia\n",
      "\tontology: https://data.bioontology.org/ontologies/HP\n",
      "Annotation details\n",
      "\tfrom: 245\n",
      "\tto: 257\n",
      "\tmatch type: PREF\n",
      "\tfrom: 313\n",
      "\tto: 325\n",
      "\tmatch type: PREF\n",
      "\tfrom: 1467\n",
      "\tto: 1479\n",
      "\tmatch type: PREF\n",
      "\tfrom: 1538\n",
      "\tto: 1550\n",
      "\tmatch type: PREF\n",
      "\n",
      "\n",
      "\n",
      "Class details\n",
      "\tid: http://purl.obolibrary.org/obo/HP_0031797\n",
      "\tprefLabel: Clinical course\n",
      "\tontology: https://data.bioontology.org/ontologies/HP\n",
      "Annotation details\n",
      "\tfrom: 542\n",
      "\tto: 556\n",
      "\tmatch type: PREF\n",
      "\n",
      "\n",
      "\n",
      "Class details\n",
      "\tid: http://purl.obolibrary.org/obo/MP_0002169\n",
      "\tprefLabel: no abnormal phenotype detected\n",
      "\tontology: https://data.bioontology.org/ontologies/MP\n",
      "Annotation details\n",
      "\tfrom: 1038\n",
      "\tto: 1043\n",
      "\tmatch type: SYN\n",
      "\n",
      "\n",
      "\n",
      "Class details\n",
      "\tid: http://purl.obolibrary.org/obo/HP_0030646\n",
      "\tprefLabel: Peripheral\n",
      "\tontology: https://data.bioontology.org/ontologies/HP\n",
      "Annotation details\n",
      "\tfrom: 1506\n",
      "\tto: 1515\n",
      "\tmatch type: PREF\n",
      "\n",
      "\n",
      "\n",
      "Dramatic structural and functional remodeling occurs in the postpartum brain for the establishment of maternal care, which is essential for the growth and development of young offspring. Glutamate and GABA signaling are critically important in modulating multiple behavioral performances. Large scale signaling changes occur in the postpartum brain, but it is still not clear to what extent the neurotransmitters glutamate and GABA change and whether the ratio of glutamate/GABA remains balanced. In this study, we examined the glutamate/GABA-glutamine cycle in the lateral septum (LS) of postpartum female mice. In postpartum females (relative to virgins), tissue levels of glutamate and GABA were elevated in LS and increased mRNA was found for the respective enzymes producing glutamate and GABA, glutaminase (Gls) and glutamate decarboxylase 1 and 2 (Gad1 and Gad2). The common precursor, glutamine, was elevated as was the enzyme that produces it, glutamate-ammonia ligase (Glul). Additionally, glutamate, GABA, and glutamine were positively correlated and the glutamate/GABA ratio was almost identical in the postpartum and virgin females. Collectively, these findings indicate that glutamate and GABA signaling are increased and that the ratio of glutamate/GABA is well balanced in the maternal LS. The postpartum brain may provide a useful model system for understanding how glutamate and GABA are linked despite large signaling changes. Given that some mental health disorders, including depression and schizophrenia display dysregulated glutamate/GABA ratio, and there is increased vulnerability to mental disorders in mothers, it is possible that these postpartum disorders emerge when glutamate and GABA changes are not properly coordinated. \n",
      "Class details\n",
      "\tid: http://purl.obolibrary.org/obo/HP_0025275\n",
      "\tprefLabel: Lateral\n",
      "\tontology: https://data.bioontology.org/ontologies/HP\n",
      "Annotation details\n",
      "\tfrom: 567\n",
      "\tto: 573\n",
      "\tmatch type: PREF\n",
      "\n",
      "\n",
      "\n",
      "Class details\n",
      "\tid: http://purl.obolibrary.org/obo/HP_0000716\n",
      "\tprefLabel: Depression\n",
      "\tontology: https://data.bioontology.org/ontologies/HP\n",
      "Annotation details\n",
      "\tfrom: 1498\n",
      "\tto: 1507\n",
      "\tmatch type: PREF\n",
      "\n",
      "\n",
      "\n",
      "Class details\n",
      "\tid: http://purl.obolibrary.org/obo/HP_0100753\n",
      "\tprefLabel: Schizophrenia\n",
      "\tontology: https://data.bioontology.org/ontologies/HP\n",
      "Annotation details\n",
      "\tfrom: 1513\n",
      "\tto: 1525\n",
      "\tmatch type: PREF\n",
      "\n",
      "\n",
      "\n",
      "Ventral tegmental area (VTA) neurons play roles in reward and aversion. We recently discovered that the VTA has neurons that co-transmit glutamate and GABA (glutamate-GABA co-transmitting neurons), transmit glutamate without GABA (glutamate-transmitting neurons), or transmit GABA without glutamate (GABA-transmitting neurons). However, the functions of these VTA cell types in motivated behavior are unclear. To identify the functions of these VTA cell types, we combine recombinase mouse lines with INTRSECT2.0 vectors to selectively target these neurons. We find that VTA cell types have unique signaling patterns for reward, aversion, and learned cues. Whereas VTA glutamate-transmitting neurons signal cues predicting reward, VTA GABA-transmitting neurons signal cues predicting the absence of reward, and glutamate-GABA co-transmitting neurons signal rewarding and aversive outcomes without signaling learned cues related to those outcomes. Thus, we demonstrate that genetically defined subclasses of VTA glutamate and GABA neurons signal different aspects of motivated behavior.\n",
      "To investigate the effect of propofol on the release of glutamate and gamma-aminobutyric acid (GABA) from rat hippocampal synatosomes, synaptosomes was made from hippocampus and incubated with artificial cerebrospinal fluid (aCSF). With the experiment of Ca(2+)-dependent release of glutamate and GABA, dihydrokainic acid (DHK) and nipectic acid were added into aCSF. For the observation of Ca(2+)-independent release of glutamate and GABA, no DHK, nipectic acid and Ca2+ were added from aCSF. The release of glutamate and GABA were evoked by 20 micromol/L veratridine or 30 mmol/L KCI. The concentration of glutamate and GABA in aCSF was measured by using high-performance liquid chromatography (HPLC). 30, 100 and 300 micromol/L propofol significantly inhibited veratridine-evoked Ca(2+)-dependent release of glutamate and GABA (P < 0.01 or P < 0. 05). However, propofol showed no effect on elevated KCl-evoked Ca(2+)-dependent release of glutamate and GABA (P > 0.05). Veratridine or elevated KCI evoked Ca(2+)-independent release of glutamate and GABA was not affected significantly by propofol (P > 0.05). Propofol could inhibit Ca(2+)-dependent release of glutamate and GABA. However, it has no effect on the Ca(2+)-independent release of glutamate and GABA.\n",
      "Class details\n",
      "\tid: http://purl.obolibrary.org/obo/HP_0032320\n",
      "\tprefLabel: Affected\n",
      "\tontology: https://data.bioontology.org/ontologies/HP\n",
      "Annotation details\n",
      "\tfrom: 1065\n",
      "\tto: 1072\n",
      "\tmatch type: PREF\n",
      "\n",
      "\n",
      "\n",
      "There is ample evidence that the inhibitory GABA and the excitatory glutamate system are essential for an adequate response to stress. Both GABAergic and glutamatergic brain circuits modulate hypothalamus-pituitary-adrenal (HPA)-axis activity, and stress in turn affects glutamate and GABA levels in the rodent brain. However, studies examining stress-induced GABA and glutamate levels in the human brain are scarce. Therefore, we investigated the influence of acute psychosocial stress (using the Trier Social Stress Test) on glutamate and GABA levels in the medial prefrontal cortex of 29 healthy male individuals using 7 Tesla proton magnetic resonance spectroscopy. <i>In vivo</i> GABA and glutamate levels were measured before and 30 min after exposure to either the stress or the control condition. We found no associations between psychosocial stress or cortisol stress reactivity and changes over time in medial prefrontal glutamate and GABA levels. GABA and glutamate levels over time were significantly correlated in the control condition but not in the stress condition, suggesting that very subtle differential effects of stress on GABA and glutamate across individuals may occur. However, overall, acute psychosocial stress does not appear to affect <i>in vivo</i> medial prefrontal GABA and glutamate levels, at least this is not detectable with current practice <sup>1</sup>H-MRS.\n",
      "Class details\n",
      "\tid: http://purl.obolibrary.org/obo/HP_0011009\n",
      "\tprefLabel: Acute\n",
      "\tontology: https://data.bioontology.org/ontologies/HP\n",
      "Annotation details\n",
      "\tfrom: 462\n",
      "\tto: 466\n",
      "\tmatch type: PREF\n",
      "\tfrom: 1212\n",
      "\tto: 1216\n",
      "\tmatch type: PREF\n",
      "\n",
      "\n",
      "\n",
      "Class details\n",
      "\tid: http://purl.obolibrary.org/obo/HP_0032322\n",
      "\tprefLabel: Healthy\n",
      "\tontology: https://data.bioontology.org/ontologies/HP\n",
      "Annotation details\n",
      "\tfrom: 592\n",
      "\tto: 598\n",
      "\tmatch type: PREF\n",
      "\n",
      "\n",
      "\n",
      "Brain microdialysis has become a frequently used method to study the extracellular concentrations of neurotransmitters in specific areas of the brain. For years, and this is still the case today, dialysate concentrations and hence extracellular concentrations of neurotransmitters have been interpreted as a direct index of the neuronal release of these specific neurotransmitter systems. Although this seems to be the case for neurotransmitters such as dopamine, serotonin and acetylcholine, the extracellular concentrations of glutamate and GABA do not provide a reliable index of their synaptic exocytotic release. However, many microdialysis studies show changes in extracellular concentrations of glutamate and GABA under specific pharmacological and behavioural stimuli that could be interpreted as a consequence of the activation of specific neurochemical circuits. Despite this, we still do not know the origin and physiological significance of these changes of glutamate and GABA in the extracellular space. Here we propose that the changes in dialysate concentrations of these two neurotransmitters found under specific treatments could be an expression of the activity of the neurone-astrocyte unit in specific circuits of the brain. It is further proposed that dialysate changes of glutamate and GABA could be used as an index of volume transmission mediated actions of these two neurotransmitters in the brain. This hypothesis is based firstly on the assumption that the activity of neurones is functionally linked to the activity of astrocytes, which can release glutamate and GABA to the extracellular space; secondly, on the existence of extrasynaptic glutamate and GABA receptors with functional properties different from those of GABA receptors located at the synapse; and thirdly, on the experimental evidence reporting specific electrophysiological and neurochemical effects of glutamate and GABA when their levels are increased in the extracellular space. According to this concept, glutamate and GABA, once released into the extracellular compartment, could diffuse and have long-lasting effects modulating glutamatergic and/or GABAergic neurone-astrocytic networks and their interactions with other neurotransmitter neurone networks in the same areas of the brain.\n",
      "Class details\n",
      "\tid: http://purl.obolibrary.org/obo/HP_0020034\n",
      "\tprefLabel: Diffuse\n",
      "\tontology: https://data.bioontology.org/ontologies/HP\n",
      "Annotation details\n",
      "\tfrom: 2081\n",
      "\tto: 2087\n",
      "\tmatch type: PREF\n",
      "\n",
      "\n",
      "\n",
      "Disturbances in the brain glutamate and GABA (γ-aminobutyric acid) homeostasis may be markers of transition to psychosis in individuals at high-risk (HR). Knowledge of GABA and glutamate levels in HR stages could give an insight into changes in the neurochemistry underlying psychosis. Studies on glutamate in HR have provided conflicting data, and GABA studies have only recently been initialized. In this meta-analysis, we compared cerebral levels of glutamate and GABA in HR individuals with healthy controls (HC). We searched Medline and Embase for articles published on <sup>1</sup>H-MRS studies on glutamate and GABA in HR states until April 9th, 2019. We identified a total of 28 eligible studies, of which eight reported GABA (243 HR, 356 HC) and 26 reported glutamate (299 HR, 279 HC) or Glx (glutamate + glutamine) (584 HR, 632 HC) levels. Sample sizes varied from 6 to 75 for HR and 10 to 184 for HC. Our meta-analysis of <sup>1</sup>H-MRS studies on glutamate and GABA in HR states displayed significantly lower (P = 0.0003) levels of thalamic glutamate in HR individuals than in HC and significantly higher (P = 0.001) Glx in the frontal lobe of genetic HR individuals (1st-degree relatives) than in HC. No other significant differences in glutamate and GABA levels were found. Subject numbers in the studies on glutamate as well as GABA levels were generally small and the data conflicting. Our meta-analytical findings highlight the need for larger and more homogeneous studies of glutamate and GABA in high-risk states.\n",
      "Class details\n",
      "\tid: http://purl.obolibrary.org/obo/HP_0000709\n",
      "\tprefLabel: Psychosis\n",
      "\tontology: https://data.bioontology.org/ontologies/HP\n",
      "Annotation details\n",
      "\tfrom: 112\n",
      "\tto: 120\n",
      "\tmatch type: PREF\n",
      "\tfrom: 276\n",
      "\tto: 284\n",
      "\tmatch type: PREF\n",
      "\n",
      "\n",
      "\n",
      "Class details\n",
      "\tid: http://purl.obolibrary.org/obo/HP_0032322\n",
      "\tprefLabel: Healthy\n",
      "\tontology: https://data.bioontology.org/ontologies/HP\n",
      "Annotation details\n",
      "\tfrom: 496\n",
      "\tto: 502\n",
      "\tmatch type: PREF\n",
      "\n",
      "\n",
      "\n",
      "Neurons are metabolically handicapped in the sense that they are not able to perform de novo synthesis of neurotransmitter glutamate and gamma-aminobutyric acid (GABA) from glucose. A metabolite shuttle known as the glutamate/GABA-glutamine cycle describes the release of neurotransmitter glutamate or GABA from neurons and subsequent uptake into astrocytes. In return, astrocytes release glutamine to be taken up into neurons for use as neurotransmitter precursor. In this review, the basic properties of the glutamate/GABA-glutamine cycle will be discussed, including aspects of transport and metabolism. Discussions of stoichiometry, the relative role of glutamate vs. GABA and pathological conditions affecting the glutamate/GABA-glutamine cycling are presented. Furthermore, a section is devoted to the accompanying ammonia homeostasis of the glutamate/GABA-glutamine cycle, examining the possible means of intercellular transfer of ammonia produced in neurons (when glutamine is deamidated to glutamate) and utilized in astrocytes (for amidation of glutamate) when the glutamate/GABA-glutamine cycle is operating. A main objective of this review is to endorse the view that the glutamate/GABA-glutamine cycle must be seen as a bi-directional transfer of not only carbon units but also nitrogen units.\n",
      "Bipolar disorder (BD) is characterized by unstable mood states ranging from mania to depression. Although there is some evidence that mood instability may result from an imbalance between excitatory glutamatergic and inhibitory GABA-ergic neurotransmission, few proton magnetic resonance spectroscopy (<sup>1</sup>H-MRS) studies have measured these two neurometabolites simultaneously in BD. The enzyme glutamic acid decarboxylase (GAD1) catalyzes the decarboxylation of glutamate (Glu) to GABA, and its single nucleotide polymorphisms (SNPs) might influence Glu/GABA ratio. Thus, we investigated Glu/GABA ratio in the dorsal anterior cingulate cortex (dACC) of euthymic BD type I patients and healthy controls (HC), and assessed the influence of both mood stabilizers and GAD1 SNPs on this ratio. Eighty-eight subjects (50 euthymic BD type I patients and 38 HC) underwent 3T <sup>1</sup>H-MRS in the dACC (2 × 2 × 4.5 cm<sup>3</sup>) using a two-dimensional JPRESS sequence and all subjects were genotyped for 4 SNPs in the GAD1 gene. BD patients had lower dACC Glu/GABA ratio compared to HC, where this was influenced by anticonvulsant and antipsychotic medications, but not lithium. The presence of GAD1 rs1978340 allele A was associated with higher Glu/GABA ratio in BD, while patients without this allele taking mood stabilizers had a lower Glu/GABA ratio. The lowering of dACC Glu/GABA could be one explanation for the mood stabilizing action of anticonvulsants and antipsychotics in BD type I euthymia. Therefore, this putative role of Glu/GABA ratio and the influence of GAD1 genotype interacting with mood stabilization medication should be confirmed by further studies involving larger samples and other mood states.ClincalTrials.gov registration: NCT01237158.\n",
      "Class details\n",
      "\tid: http://purl.obolibrary.org/obo/HP_0007302\n",
      "\tprefLabel: Bipolar affective disorder\n",
      "\tontology: https://data.bioontology.org/ontologies/HP\n",
      "Annotation details\n",
      "\tfrom: 1\n",
      "\tto: 16\n",
      "\tmatch type: SYN\n",
      "\n",
      "\n",
      "\n",
      "Class details\n",
      "\tid: http://purl.obolibrary.org/obo/HP_0100754\n",
      "\tprefLabel: Mania\n",
      "\tontology: https://data.bioontology.org/ontologies/HP\n",
      "Annotation details\n",
      "\tfrom: 77\n",
      "\tto: 81\n",
      "\tmatch type: PREF\n",
      "\n",
      "\n",
      "\n",
      "Class details\n",
      "\tid: http://purl.obolibrary.org/obo/HP_0000716\n",
      "\tprefLabel: Depression\n",
      "\tontology: https://data.bioontology.org/ontologies/HP\n",
      "Annotation details\n",
      "\tfrom: 86\n",
      "\tto: 95\n",
      "\tmatch type: PREF\n",
      "\n",
      "\n",
      "\n",
      "Class details\n",
      "\tid: http://purl.obolibrary.org/obo/HP_0002172\n",
      "\tprefLabel: Postural instability\n",
      "\tontology: https://data.bioontology.org/ontologies/HP\n",
      "Annotation details\n",
      "\tfrom: 171\n",
      "\tto: 179\n",
      "\tmatch type: SYN\n",
      "\n",
      "\n",
      "\n",
      "Class details\n",
      "\tid: http://purl.obolibrary.org/obo/HP_0032322\n",
      "\tprefLabel: Healthy\n",
      "\tontology: https://data.bioontology.org/ontologies/HP\n",
      "Annotation details\n",
      "\tfrom: 695\n",
      "\tto: 701\n",
      "\tmatch type: PREF\n",
      "\n",
      "\n",
      "\n",
      "Magnetic resonance spectroscopy (MRS) measures the two most common inhibitory and excitatory neurotransmitters, GABA and glutamate, in the human brain. However, the role of MRS-derived GABA and glutamate signals in relation to system-level neural signaling and behavior is not fully understood. In this study, we investigated levels of GABA and glutamate in the visual cortex of healthy human participants (both genders) in three functional states with increasing visual input. Compared with a baseline state of eyes closed, GABA levels decreased after opening the eyes in darkness and Glx levels remained stable during eyes open but increased with visual stimulation. In relevant states, GABA and Glx correlated with amplitude of fMRI signal fluctuations. Furthermore, visual discriminatory performance correlated with the level of GABA, but not Glx. Our study suggests that differences in brain states can be detected through the contrasting dynamics of GABA and Glx, which has implications in interpreting MRS measurements.<b>SIGNIFICANCE STATEMENT</b> GABA and glutamate are the two most abundant neurotransmitters in human brain. Their interaction, known as inhibitory-excitatory balance, plays a crucial role in establishing spontaneous and stimulus-driven brain activity. Yet, the relationship between magnetic resonance spectroscopy (MRS)-derived levels of both metabolites and fMRI is still a matter of dispute. In this work, we study GABA and glutamate in three states of visual processing and in relation to fMRI and visual discriminatory performance in healthy people. We found that states of visual processing can be detected through the contrasting dynamics of GABA and glutamate and their correlation with fMRI signals. We also demonstrated that GABA, but not glutamate, in the visual system predicts visual performance. Our results provide insights into MRS-derived GABA and glutamate measurements.\n",
      "Class details\n",
      "\tid: http://purl.obolibrary.org/obo/HP_0032322\n",
      "\tprefLabel: Healthy\n",
      "\tontology: https://data.bioontology.org/ontologies/HP\n",
      "Annotation details\n",
      "\tfrom: 380\n",
      "\tto: 386\n",
      "\tmatch type: PREF\n",
      "\tfrom: 1566\n",
      "\tto: 1572\n",
      "\tmatch type: PREF\n",
      "\n",
      "\n",
      "\n",
      "Class details\n",
      "\tid: http://purl.obolibrary.org/obo/HP_0031915\n",
      "\tprefLabel: Stable\n",
      "\tontology: https://data.bioontology.org/ontologies/HP\n",
      "Annotation details\n",
      "\tfrom: 607\n",
      "\tto: 612\n",
      "\tmatch type: PREF\n",
      "\n",
      "\n",
      "\n",
      "In patients suffering from stress-related pathologies and depression, frontal cortex GABA and glutamate contents are reported to decrease and increase, respectively. This suggests that the GABA and/or glutamate content may participate in pathological phenotype expression. Whether differences in frontal cortex GABA and glutamate contents would be associated with specific behavioral and neurobiological patterns remains unclear, especially in the event of exposure to moderate stress. We hypothesized that an increase in prefrontal cortex GABA/glutamate ratio would be associated with a blunted prefrontal cortex activation, an enhanced hypothalamo-pituitary-adrenocortical (HPA) axis activation and changes in behavior. Rats being restrained for 1-h were then tested in an open-field test in order to assess their behavior while under stress, and were sacrificed immediately afterward. The GABA/glutamate ratio was assessed by (1)H high-resolution magic angle spinning magnetic resonance spectroscopy ((1)H-HRMAS-MRS). The neurobiological response was evaluated through prefrontal cortex mRNA expression and plasma corticosterone levels. The stressed rats were distributed into two subgroups according to their high (H-G/g) or low (L-G/g) GABA/glutamate ratio. Compared to the L-G/g rats, the H-G/g rats exhibited a decrease in c-fos, Arc, Npas4, Nr4a2 mRNA expression suggesting blunted prefrontal cortex activation. They also showed a more pronounced stress with an enhanced rise in corticosterone, alanine aminotransferase (ALAT), aspartate aminotransferase (ASAT), creatine kinase (CK) and lactate dehydrogenase (LDH) levels, as well as behavioral disturbances with decreased locomotion speed. These changes were independent from prefrontal cortex energetic status as mammalian target of rapamycin (mTOR) and adenosine monophosphate-activated protein kinase (AMPK) pathway activities were similar in both subpopulations. The differences in GABA/glutamate ratio in the frontal cortex observed in the stressed animals may participate in shaping individual differences in psychophysiological reactions. \n",
      "Class details\n",
      "\tid: http://purl.obolibrary.org/obo/HP_0000716\n",
      "\tprefLabel: Depression\n",
      "\tontology: https://data.bioontology.org/ontologies/HP\n",
      "Annotation details\n",
      "\tfrom: 59\n",
      "\tto: 68\n",
      "\tmatch type: PREF\n",
      "\n",
      "\n",
      "\n",
      "Class details\n",
      "\tid: http://purl.obolibrary.org/obo/HP_0012826\n",
      "\tprefLabel: Moderate\n",
      "\tontology: https://data.bioontology.org/ontologies/HP\n",
      "Annotation details\n",
      "\tfrom: 470\n",
      "\tto: 477\n",
      "\tmatch type: PREF\n",
      "\n",
      "\n",
      "\n",
      "Class details\n",
      "\tid: http://purl.obolibrary.org/obo/MP_0001411\n",
      "\tprefLabel: spinning\n",
      "\tontology: https://data.bioontology.org/ontologies/MP\n",
      "Annotation details\n",
      "\tfrom: 963\n",
      "\tto: 970\n",
      "\tmatch type: PREF\n",
      "\n",
      "\n",
      "\n",
      "Class details\n",
      "\tid: http://purl.obolibrary.org/obo/HP_0000708\n",
      "\tprefLabel: Behavioral abnormality\n",
      "\tontology: https://data.bioontology.org/ontologies/HP\n",
      "Annotation details\n",
      "\tfrom: 1644\n",
      "\tto: 1666\n",
      "\tmatch type: SYN\n",
      "\n",
      "\n",
      "\n",
      "Three of twelve secoergoline derivatives (Z ethyl 4-[(ethoxycarbonylmethyl)methylamino]-2-methyl-3-phenylpent-2-enoate, 8; ethyl 1,6-dimethyl-3-oxo-5-phenyl-1,2,3,6-tetrahydropyridine-2-carboxylate, 9; Z methyl 4-[(methoxycarbonylmethyl)methylamino)-2-methyl-3-phenylpent-2-enoate, 11), containing bioisosteric sequences of GABA and Glu, inhibited both GABA and Glu transport through cerebrocortical membranes specifically. Compounds 8, 9, and 11 appeared to be equipotent inhibitors of GABA and Glu transport with IC50 values between 270 and 1100 microM, whereas derivatives 1-7, 10, and 12 were without effects. In the presence of GABA and Glu transport-specific nontransportable inhibitors, inhibition of GABA and Glu transport by 8, 9, and 11 proceeded in two phases. The two phases of inhibition were characterized by IC50 values between 4 and 180 nM and 360-1020 microM and different selectivity sequences. These findings may indicate the existence of some mechanism possibly mediated by a previously unrecognized GABA-Glu transporter. Derivatives with the cis, but not the trans configuration of bulky ester groups (8 vs 7 and 11 vs 12) showed significant inhibitory effect (IC50 values of 270 microM vs >>1000 microM and 1100 microM vs >>1000 microM on GABA transport, respectively). The cis-trans selectivity can be explained by docking these secoergolines in a three-dimensional model of the second and third transmembrane helices of GABA transporter type 1.\n",
      "Iminodipropionitrile (IDPN)-induced dyskinetic syndrome is characterized by abnormal repetitive involuntary movements with abnormalities in the neuro-transmission. This study explored the mechanism of glutamate (Glu)/γ-aminobutyric acid (GABA)-glutamine (Gln) metabolic circuit in rat dyskinetic syndrome and the possible regulation mechanism of \"tiapride (Tia).\"Male Wistar rats were assigned to the control group, dyskinetic syndrome group, and Tia group. Dyskinetic syndrome was induced by injecting with 3,3'-iminodipropionitrile for 7 days. Tia group was treated with tiapride, while the control and dyskinetic syndrome groups were gavaged with saline. Eventually the Glu, GABA, and Gln concentrations in striatum were detected using UPLC-3QMS, additionally another amino acid neurotransmitters (aspartate, glycine) were also detected. Expressions of glutamine synthetase (GS), glutamate transporter (EAAT<sub>2</sub>), glutamate decarboxylase (GAD65/67), and γ-aminobutyric acid transporter protein (GAD-T) were observed using Western blot and real-time polymerase chain reaction.The behavior test scores of dyskinetic syndrome group were increased compared with the control group. Tia group decreased the behavior test scores compared with dyskinetic syndrome group. For amino acid neuro-transmission, dyskinetic syndrome group increased Glu level (p < 0.01), decreased GABA level (p < 0.01), increased Glu/GABA ratio (p < 0.01), and decreased Asp level (p < 0.01) compared with control group. Tia group decreased Glu level (p < 0.01), increased GABA level (p < 0.01), decreased Glu/GABA ratio (p < 0.01), and increased Asp level (p < 0.05) compared with dyskinetic syndrome group. For Glu/GABA-Gln circuit, the protein and mRNA expression of GS and EAAT<sub>2</sub> in dyskinetic syndrome group were decreased (p < 0.05). Tia group increased protein and mRNA expression level of GS (p < 0.05) and EAAT<sub>2</sub> (p < 0.01).The rat dyskinetic syndrome has Glu/GABA-Gln abnormalities. \"Tiapride\" upregulated the protein expression of GS and EAAT<sub>2</sub>, reduce Glu levels, increase γ-GABA levels, and eventually improve amino acid neurotransmitter imbalance.\n",
      "Class details\n",
      "\tid: http://purl.obolibrary.org/obo/HP_0004305\n",
      "\tprefLabel: Involuntary movements\n",
      "\tontology: https://data.bioontology.org/ontologies/HP\n",
      "Annotation details\n",
      "\tfrom: 97\n",
      "\tto: 117\n",
      "\tmatch type: PREF\n",
      "\n",
      "\n",
      "\n",
      "Class details\n",
      "\tid: http://purl.obolibrary.org/obo/HP_0002326\n",
      "\tprefLabel: Transient ischemic attack\n",
      "\tontology: https://data.bioontology.org/ontologies/HP\n",
      "Annotation details\n",
      "\tfrom: 358\n",
      "\tto: 360\n",
      "\tmatch type: SYN\n",
      "\tfrom: 448\n",
      "\tto: 450\n",
      "\tmatch type: SYN\n",
      "\tfrom: 547\n",
      "\tto: 549\n",
      "\tmatch type: SYN\n",
      "\tfrom: 1189\n",
      "\tto: 1191\n",
      "\tmatch type: SYN\n",
      "\tfrom: 1502\n",
      "\tto: 1504\n",
      "\tmatch type: SYN\n",
      "\tfrom: 1831\n",
      "\tto: 1833\n",
      "\tmatch type: SYN\n",
      "\n",
      "\n",
      "\n",
      "Class details\n",
      "\tid: http://purl.obolibrary.org/obo/MP_0030733\n",
      "\tprefLabel: decreased gamma-aminobutyric acid level\n",
      "\tontology: https://data.bioontology.org/ontologies/MP\n",
      "Annotation details\n",
      "\tfrom: 1368\n",
      "\tto: 1387\n",
      "\tmatch type: SYN\n",
      "\n",
      "\n",
      "\n",
      "Class details\n",
      "\tid: http://purl.obolibrary.org/obo/MP_0030732\n",
      "\tprefLabel: increased gamma-aminobutyric acid level\n",
      "\tontology: https://data.bioontology.org/ontologies/MP\n",
      "Annotation details\n",
      "\tfrom: 1544\n",
      "\tto: 1563\n",
      "\tmatch type: SYN\n",
      "\n",
      "\n",
      "\n",
      "Class details\n",
      "\tid: http://purl.obolibrary.org/obo/HP_0002172\n",
      "\tprefLabel: Postural instability\n",
      "\tontology: https://data.bioontology.org/ontologies/HP\n",
      "Annotation details\n",
      "\tfrom: 2162\n",
      "\tto: 2170\n",
      "\tmatch type: SYN\n",
      "\n",
      "\n",
      "\n",
      "Animal models of mild traumatic brain injury (mTBI) suggest that metabolic changes in the brain occur immediately after a mechanical injury to the head. Proton magnetic resonance spectroscopy (<sup>1</sup>H-MRS) can be used to determine relative concentrations of metabolites in vivo in the human brain. The purpose of this study was to determine concentrations of glutamate and GABA in the brain acutely after mTBI and throughout 2 mo of recovery. Concentrations of glutamate and GABA were obtained using <sup>1</sup>H-MRS in nine individuals who had suffered an mTBI and nine control individuals in two brain regions of interest: the primary motor cortex (M1), and the dorsolateral prefrontal cortex (DLPFC), and at three different time points postinjury: 72 h, 2 wk, and 2 mo postinjury. There were no differences between groups in concentrations of glutamate or GABA, or the ratio of glutamate to GABA, in M1. In the DLPFC, glutamate concentration was lower in the mTBI group compared with controls at 72 h postinjury (d = 1.02), and GABA concentration was lower in the mTBI group at 72 h and 2 wk postinjury (d = 0.81 and d = 1.21, respectively). The ratio of glutamate to GABA in the DLPFC was higher in the mTBI group at 2 wk postinjury (d = 1.63). These results suggest that changes in glutamate and GABA concentrations in the brain may be region-specific and may depend on the amount of time that has elapsed postinjury. NEW & NOTEWORTHY To our knowledge, this is the first study to examine neurotransmitter concentrations in vivo at multiple time points throughout recovery from mild traumatic brain injury in humans.\n",
      "Class details\n",
      "\tid: http://purl.obolibrary.org/obo/HP_0012825\n",
      "\tprefLabel: Mild\n",
      "\tontology: https://data.bioontology.org/ontologies/HP\n",
      "Annotation details\n",
      "\tfrom: 18\n",
      "\tto: 21\n",
      "\tmatch type: PREF\n",
      "\tfrom: 1590\n",
      "\tto: 1593\n",
      "\tmatch type: PREF\n",
      "\n",
      "\n",
      "\n",
      "Class details\n",
      "\tid: http://purl.obolibrary.org/obo/MP_0012448\n",
      "\tprefLabel: abnormal primary motor cortex morphology\n",
      "\tontology: https://data.bioontology.org/ontologies/MP\n",
      "Annotation details\n",
      "\tfrom: 645\n",
      "\tto: 656\n",
      "\tmatch type: SYN\n",
      "\n",
      "\n",
      "\n",
      "We developed a methodology for rapid quantification of extracellular neurotransmitters in mouse brain by PESI/MS/MS and longitudinal data analysis using the R and Stan-based Bayesian state-space model. We performed a rapid analysis for quantifying extracellular l-glutamic acid (L-Glu) and gamma-aminobutyric acid (GABA) in the mouse striatum by combined use of probe electrospray ionization/tandem mass spectrometry (PESI/MS/MS) and in vivo brain microdialysis. We optimized the PESI/MS/MS parameters with the authentic L-Glu, GABA, L-Glu-<sup>13</sup>C<sub>5</sub>,<sup>15</sup>N<sub>1</sub>, and GABA-D<sub>6</sub> standards. We constructed calibration curves of L-Glu and GABA with the stable isotope internal standard correction method (L-Glu-<sup>13</sup>C<sub>5</sub>,<sup>15</sup>N<sub>1</sub>, and GABA-D<sub>6</sub>), demonstrating sufficient linearity (R > 0.999). Additionally, the quantitative method for L-Glu and GABA was validated with low-, middle-, and high-quality control samples. The intra- and inter-day accuracy and precision were 0.4%-7.5% and 1.7%-5.4% for L-Glu, respectively, and 0.1%-4.8% and 2.1%-5.7% for GABA, respectively, demonstrating high reproducibility of the method. To evaluate the feasibility of this method, microdialyses were performed on free-moving mice that were stimulated by high-K<sup>+</sup>-induced depolarization under different sampling conditions: 1) every 5 min for 150 min (n = 2) and 2) every 1 min for 30 min (n = 3). We applied the R and Stan-based Bayesian state-space model to each mouse's time-series data considering autocorrelation, and the model successfully detected abnormal changes in the L-Glu and GABA levels in each mouse. Thus, the L-Glu and GABA levels in all microdialysates approximately increased up to two- and seven-fold levels through high-K<sup>+</sup>-induced depolarization. Additionally, a 1-min temporal resolution was achieved using this method, thereby successfully monitoring microenvironmental changes in the extracellular L-Glu and GABA of the mouse striatum. In conclusion, this methodology using PESI/MS/MS and Bayesian state-space model allowed easy monitoring of neurotransmitters at high temporal resolutions and appropriate data interpretation considering autocorrelation of time-series data, which will reveal hidden pathological mechanisms of brain diseases, such as Parkinson's disease and Huntington's disease in the future.\n",
      "Class details\n",
      "\tid: http://purl.obolibrary.org/obo/HP_0031915\n",
      "\tprefLabel: Stable\n",
      "\tontology: https://data.bioontology.org/ontologies/HP\n",
      "Annotation details\n",
      "\tfrom: 691\n",
      "\tto: 696\n",
      "\tmatch type: PREF\n",
      "\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "def hpo_extraction():\n",
    "    for abstract in rankings[\"Abstract\"]:\n",
    "        annotations = get_json(REST_URL + \"/annotator?text=\" + urllib.parse.quote(abstract) + \"&ontologies=HP,MP\")\n",
    "        \n",
    "        print(abstract)\n",
    "        print_annotations(annotations)\n",
    "hpo_extraction()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "52342e35-ccdc-4bd5-9f1b-7b73b45453e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def print_annotations(annotations, get_class=True):\n",
    "    for result in annotations:\n",
    "        class_details = result[\"annotatedClass\"]\n",
    "        if get_class:\n",
    "            try:\n",
    "                class_details = get_json(result[\"annotatedClass\"][\"links\"][\"self\"])\n",
    "            except urllib.error.HTTPError:\n",
    "                print(f\"Error retrieving {result['annotatedClass']['@id']}\")\n",
    "                continue\n",
    "        print(\"Class details\")\n",
    "        print(\"\\tid: \" + class_details[\"@id\"])\n",
    "        print(\"\\tprefLabel: \" + class_details[\"prefLabel\"])\n",
    "        print(\"\\tontology: \" + class_details[\"links\"][\"ontology\"])\n",
    "\n",
    "        print(\"Annotation details\")\n",
    "        for annotation in result[\"annotations\"]:\n",
    "            print(\"\\tfrom: \" + str(annotation[\"from\"]))\n",
    "            print(\"\\tto: \" + str(annotation[\"to\"]))\n",
    "            print(\"\\tmatch type: \" + annotation[\"matchType\"])\n",
    "\n",
    "        if result[\"hierarchy\"]:\n",
    "            print(\"\\n\\tHierarchy annotations\")\n",
    "            for annotation in result[\"hierarchy\"]:\n",
    "                try:\n",
    "                    class_details = get_json(annotation[\"annotatedClass\"][\"links\"][\"self\"])\n",
    "                except urllib.error.HTTPError:\n",
    "                    print(f\"Error retrieving {annotation['annotatedClass']['@id']}\")\n",
    "                    continue\n",
    "                pref_label = class_details[\"prefLabel\"] or \"no label\"\n",
    "                print(\"\\t\\tClass details\")\n",
    "                print(\"\\t\\t\\tid: \" + class_details[\"@id\"])\n",
    "                print(\"\\t\\t\\tprefLabel: \" + class_details[\"prefLabel\"])\n",
    "                print(\"\\t\\t\\tontology: \" + class_details[\"links\"][\"ontology\"])\n",
    "                print(\"\\t\\t\\tdistance from originally annotated class: \" + str(annotation[\"distance\"]))\n",
    "\n",
    "        print(\"\\n\\n\")\n",
    "        \n",
    "print_annotations(annotations)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:DIS] *",
   "language": "python",
   "name": "conda-env-DIS-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
